src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -6133,10 +6135,35 @@ void StorageReplicatedMergeTree::clearOldPartsAndRemoveFromZK()
     /// Delete duplicate parts from filesystem
     if (!parts_to_delete_only_from_filesystem.empty())
     {
-        clearPartsFromFilesystem(parts_to_delete_only_from_filesystem);
-        removePartsFinally(parts_to_delete_only_from_filesystem);
+        /// It can happen that some error appear during part removal from FS.
+        /// In case of such exception we have to change state of failed parts from Deleting to Outdated.
+        /// Otherwise nobody will try to remove them again (see grabOldParts).
+        NameSet parts_failed_to_delete;
+        tryClearPartsFromFilesystem(parts_to_delete_only_from_filesystem, parts_failed_to_delete);
 
-        LOG_DEBUG(log, "Removed {} old duplicate parts", parts_to_delete_only_from_filesystem.size());
+        DataPartsVector finally_remove_parts;
+        if (!parts_failed_to_delete.empty())
+        {
+            DataPartsVector rollback_parts;
+            for (const auto & part : parts_to_delete_only_from_filesystem)
+            {
+                if (!parts_failed_to_delete.contains(part->name))
+                    finally_remove_parts.push_back(part);
+                else
+                    rollback_parts.push_back(part);
+            }
+
+            if (!rollback_parts.empty())
+                rollbackDeletingParts(rollback_parts);
```

Comment:
I've suddenly realized that this method is dangerous because of this: https://github.com/ClickHouse/ClickHouse/issues/36610#issuecomment-1117517863. Similar issue may happen if part `all_2_2_0` will become `Deleting` before execution of `DROP_RANGE`, but then will be rolled back to `Outdated`. So `DROP_RANGE` will be successfully processed, but a covered part will not be removed from zk.

Reply:
Improved according to chat